kubernetes/cluster-dump-restore.md

```diff
@@ -0,0 +1,93 @@
+
+This migration process boils down to selectively omitting state that is not portable between clusters, as well as making some assumptions about the clusters we're dealing with. Those assumptions include:
+* **kube-dns add-on** This allows abstracting away the services' clusterIPs, which is inherentely non-portable across clusters with different serviceCIDRS.
+
+* **api object compatibilty** Any api object type (eg: replicaset) that is dumped from the *source cluster* must be supported by the *target cluster*.
+
+	As a general rule, if the *target cluster's* Kubernetes version is >= *source cluster's* Kubernetes version, this will not be a problem.
+
+* **blank target slate** The *target cluster* to has no deployment state beyond a "stock" cluster. In the CoreOS case, the stock cluster state is everything in the `kube-system` namespace along with the service token in the default namespace.
+
+	If the *target cluster* has other pre-existing state, more precautions must be taken to ensure there are no undesired interactions between pre-existing resources on the *target cluster* and the dump coming from *source cluster*.
+
+	Name conflicts are easy to defect, but more subtle issues like services selecting unintended pre-existing pods is another more subtle situation that can arise. We will assume **blank target slate** for the remainder of this document.
+
+As of now, this is not entirely supported upstream. This `get --export` flag is supposed to do something like this, but leaves plenty of state that we do not consider to be portable in this case. [github issue](https://github.com/kubernetes/kubernetes/issues/21582).
+
+We'll explain step-by-step how the CoreOS infrastructure team migrates deployments between Kubernetes clusters today.
+
+### Dump from the source cluster
+
+Create a directory to hold your dump files:
+
+```shell
+mkdir ./cluster-dump
+```
+
+
+First, get a list of all namespaces that are not `kube-system` or `default` and record them to a file on disk. This represents the list of namespaces that we want to migrate:
+
+```shell
+kubectl get --export -o=json ns | \
+jq '.items[] |
+	select(.metadata.name!="kube-system") |
+	select(.metadata.name!="default") |
+	del(.status,
+        .metadata.uid,
+        .metadata.selfLink,
+        .metadata.resourceVersion,
+        .metadata.creationTimestamp,
+        .metadata.generation
+    )' > ./cluster-dump/ns.json
+```
+
+
+For each of these namespaces, dump all services, replication controllers, secrets and daemonsets to a file on disk. Strip any non-portable fields from the objects. If you wish to migrate additional controller resource types (replicasets, deployments, etc), make sure to add them to the resource type list:
```

Comment:
Unless there's a good reason you wouldn't want to restore these things, we should just add them to the list in the code and in the opening sentence and delete the "if you wish..." stuff.